feat(kimi_linear): context parallelism and packed-sequence support (#3243)
NVIDIA-NeMo/Automodel: docs/model-coverage/llm/moonshotai/kimi-linear.mdx

---
title: "Kimi Linear"
description: ""
---

> [Kimi Linear](https://huggingface.co/moonshotai/Kimi-Linear-48B-A3B-Instruct) is a hybrid-attention Mixture-of-Experts language model from Moonshot AI. Most layers use Kimi Delta Attention (KDA), a gated linear-attention variant with a recurrent state, and the remaining layers use full Multi-head Latent Attention (MLA). NeMo AutoModel ships a native `KimiLinearForCausalLM` implementation with expert parallelism, packed sequences, and context parallelism.

Reviewer: ```suggestion
[Kimi Linear](https://huggingface.co/moonshotai/Kimi-Linear-48B-A3B-Instruct) is a hybrid-attention Mixture-of-Experts language model from Moonshot AI. Most layers use Kimi Delta Attention (KDA), a gated linear-attention variant with a recurrent state, and the remaining layers use full Multi-Head Latent Attention (MLA). NeMo AutoModel ships a native `KimiLinearForCausalLM` implementation with expert parallelism, packed sequences, and context parallelism.
```

<Info>

> 

Reviewer: Delete blank line after opening tag

| | |
|---|---|
| **Task** | Text Generation (hybrid linear attention, MoE) |
| **Architecture** | `KimiLinearForCausalLM` |
| **Parameters** | 48B total / 3B active |
| **HF Org** | [moonshotai](https://huggingface.co/moonshotai) |

</Info>

## Available Models

- **Kimi-Linear-48B-A3B-Instruct**: 48B total, 3B activated

## Architecture

- `KimiLinearForCausalLM` (`model_type: kimi_linear`)
- Hybrid attention stack: KDA linear-attention layers interleaved with MLA layers
- MoE feed-forward blocks with sigmoid routing and grouped top-k selection

## Supported Parallelism

| Feature | Notes |
|---|---|
| FSDP2 | Default sharding strategy for the recipes below |
| Expert parallelism (EP) | Shards the MoE experts across ranks |

> | Context parallelism (CP) | Contiguous per-rank sequence shards; KDA passes its recurrent state rank to rank, MLA gathers the compressed KV latent |

Reviewer: ```suggestion
| Context parallelism (CP) | Contiguous per-rank sequence shards. KDA passes its recurrent state from rank to rank, and MLA gathers the compressed KV latent |
```

| Packed sequences | Document boundaries from the THD collater are honored by both layer types, including under CP |

> ## Example HF Models

Reviewer: ```suggestion
## Example Hugging Face Models
```

| Model | HF ID |
|---|---|
| Kimi Linear 48B A3B Instruct | [`moonshotai/Kimi-Linear-48B-A3B-Instruct`](https://huggingface.co/moonshotai/Kimi-Linear-48B-A3B-Instruct) |

## Example Recipes

| Recipe | Description |
|---|---|
| [kimi_linear_48b_a3b_hellaswag.yaml](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/kimi/kimi_linear_48b_a3b_hellaswag.yaml) | SFT: Kimi Linear 48B A3B on HellaSwag with EP=8 |
| [kimi_linear_48b_a3b_longcontext_cp2.yaml](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/kimi/kimi_linear_48b_a3b_longcontext_cp2.yaml) | SFT: 32k packed sequences with CP=2 and EP=8 |

## Try with NeMo AutoModel

> **1. Clone and install from source** ([full instructions](/get-started/installation)):

Reviewer: ```suggestion
1. Clone and install from source. For full instructions, refer to [Installation](/get-started/installation):
```

```bash
git clone https://github.com/NVIDIA-NeMo/Automodel.git
cd Automodel
uv sync --locked --all-groups --all-extras
```

> **2. Run the recipe** from inside the repo:

Reviewer: ```suggestion
2. Run the recipe from inside the repository:
```

```bash
uv run automodel --nproc-per-node=8 examples/llm_finetune/kimi/kimi_linear_48b_a3b_hellaswag.yaml
```

<Accordion title="Run with Docker">

> **1. Pull the container** and mount a checkpoint directory:

Reviewer: ```suggestion
1. Pull the container and mount a checkpoint directory:
```

```bash
docker run --gpus all -it --rm \
  --shm-size=8g \
  -v $(pwd)/checkpoints:/opt/Automodel/checkpoints \
  nvcr.io/nvidia/nemo-automodel:26.06.00
```

> **2.** Navigate to the AutoModel directory (where the recipes are):

Reviewer: ```suggestion
2. Navigate to the AutoModel directory where the recipes are located:
```

```bash
cd /opt/Automodel
```

> **3. Run the recipe**:

Reviewer: ```suggestion
3. Run the recipe:
```

```bash
automodel --nproc-per-node=8 examples/llm_finetune/kimi/kimi_linear_48b_a3b_hellaswag.yaml
```
</Accordion>

> See the [Installation Guide](/get-started/installation) and [LLM Fine-Tuning Guide](/recipes-e2e-examples/sft-peft).

Reviewer: ```suggestion
Refer to the [Installation Guide](/get-started/installation) and the [LLM Fine-Tuning Guide](/recipes-e2e-examples/sft-peft).
```

## Fine-Tuning

> See the [LLM Fine-Tuning Guide](/recipes-e2e-examples/sft-peft).

Reviewer: ```suggestion
Refer to the [LLM Fine-Tuning Guide](/recipes-e2e-examples/sft-peft).
```
